server/backend/api/app/routers/router_guards.py

```diff
@@ -184,31 +202,69 @@ def authorize_by_application_role(
+    This authorize_by_application_role method is used for the authorization check of a specific application,
+    it is the same and depends on the method authorize_by_app_id() but for
+    the need that some routers contains target role_id in the request (instead of application_id)
+    we need to get application_id from role and check if role exists first, and then call authorize_by_app_id()
     """
-    if not role:
-        raise HTTPException(
-            status_code=HTTPStatus.FORBIDDEN,
-            detail={
-                "code": ERROR_INVALID_ROLE_ID,
-                "description": f"Requester has no appropriate role.",
-            },
-            headers={"WWW-Authenticate": "Bearer"},
-        )
-
     # Delegate to this for main check.
     authorize_by_app_id(
         application_id=role.application_id, db=db, claims=claims, requester=requester
     )
     return role
 
 
-async def internal_only_action(requester: Requester = Depends(get_current_requester)):
-    if requester.user_type_code is not UserType.IDIR:
-        raise external_user_prohibited_exception
+async def authorize_by_privilege(
+    request: Request,
+    role: FamRole = Depends(get_request_role_from_id),
+    db: Session = Depends(database.get_db),
+    claims: dict = Depends(validate_token),
+    requester: Requester = Depends(get_current_requester),
+):
+    """
+    This authorize_by_privilege method is used for checking if the requester has the privilege to grant/remove access of the role.
+    :param role: for remove access, it is the concrete role get by user_role_xref_id in the request params;
+                 for grant access, it is the concrete role, or the parent role of an abstract role get by role_id in the request params
+    """
+    requester_is_app_admin = is_app_admin(role.application_id, db, claims)
+
+    # if requester is not application admin, check if has the privilege to grant/remove access of the role
+    if not requester_is_app_admin:
+        # for remove access, role is what we get above from Depends(get_request_role_from_id)
+        # for grant access, if the request params has forest_client_number, need to get the child role
+        # the child role should already exist when the delegated admin has been created
+        if "user_role_xref_id" not in request.path_params:
+            rbody = await request.json()
+            forest_client_number = rbody.get("forest_client_number")
+            if (
+                forest_client_number
+                and role.role_type_code == RoleType.ROLE_TYPE_ABSTRACT
+            ):
+                parent_role = role
+                forest_client_role_name = (
+                    crud_user_role.construct_forest_client_role_name(
+                        parent_role.role_name, forest_client_number
+                    )
+                )
+                role = crud_role.get_role_by_role_name_and_app_id(
```

Comment:
I put the comment in the code, also want to mention here, so for granting access, it can be granting an abstract role with a forest client number, for example, the request data could be as the following. But when we check for privilege, we want to check if the requester has the privilege to grant the child role, which is "FOM_SUBMITTER_00001011", not the parent role "FOM_SUBMITTER". The code above is just to grab the child role for this special case. 
``` 
{
user_name: 'IANLIU',
user_type_code: 'I',
role_id: 3,
forest_client_number: '00001011'
}
```

Reply:
Thanks Catherine for implement this guard. I admit I was confused. This guard method is probably most difficult so it appears is quite complicated for now.
If it works (I believe), we probably later need to refactor to abstract some logic out for more clarity for future developers.